lib/Sema/MiscDiagnostics.cpp

```diff
@@ -2995,8 +2996,14 @@ static void checkStmtConditionTrailingClosure(TypeChecker &TC, const Expr *E) {
       // Dig into implicit expression.
       if (E->isImplicit()) return { true, E };
       // Diagnose call expression.
-      if (auto CE = dyn_cast<CallExpr>(E))
+      // Walk into call expressions to diagnose call chains that may have
+      // trailing closures somewhere in the middle
+      if (auto CE = dyn_cast<CallExpr>(E)) {
         diagnoseIt(CE);
+        return { true, E };
+      } else if (isa<DotSyntaxCallExpr>(E)) {
+        return { true, E };
+      }
```

Comment:
To handle binary expressions et al., I think we should maintain "not walk into" list, instead of walk into specific expression kinds.
I mean, currently we walk into only `CallExpr` (and `DotSyntaxCallExpr`). But I think we should walk into *except* `ParenExpr`, `TupleExpr`, `ArrayExpr`, `DictionaryExpr`, etc.

Reply:
I ended up adopting a solution like this, my one concern is that a future new expression kind might cause a false positive to be diagnosed if it isn't included in the 'skipped' list. This should be mitigated by the fact that isValidTrailingClosure() still whitelists the individual tokens which may follow a closing brace for something to be considered a possibly valid trailing closure. It doesn't completely eliminate the possibility of future breakage though.

Reply:
I think it's OK. False positive is better than false negative in this case. The user can easily workaround it by not using trailing closure.

Reply:
…as long as we don't break source compat.